### PR TITLE
add evaluation code

### PR DESCRIPTION
**Pull request by rrevoid**

### CODE REVIEW

File: scripts/evaluate_model.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+import numpy as np
+
+import torch
+from torch.utils.data import Dataset, DataLoader
+from torchtext.data.metrics import bleu_score
+
+from tqdm import tqdm
+
+from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
```
Comment (shmpanski) [inline scripts/evaluate_model.py]: Для чистоты кода хорошо использовать `isort` или похожие утилиты из `pycharm` / `vscode`
В `pycharm` есть [целый раздел](https://www.jetbrains.com/help/pycharm/creating-and-optimizing-imports.html#optimize-imports) про оптимизацию импортов

File: scripts/evaluate_model.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+class ConvAI2CandidatesDataset(Dataset):
+    """ConvAI2 Candidates Dataset"""
+
+    def __init__(self, dataset):
+        """
+        Args:
+            dataset (DatasetDict): ConvAI2 dataset from huggingface
+        """
+
+        np.random.seed(42)
```
Comment (shmpanski) [inline scripts/evaluate_model.py]: Есть ли необходимость хардкодить здесь `seed`?
Возможно, стоит вынести его в аргументы или отдельную функцию "`seed_all`"

File: scripts/evaluate_model.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        return sample
+
+
+def calc_metrics(generated_batch, candidates_batch, history):
+    for i, candidates in enumerate(zip(*candidates_batch)):
+        bleus = []
+        generated = list(generated_batch[i].split())
+
+        for candidate in candidates:
+            f1s.append(bleu_score([generated], [candidate], max_n=1, weights=[1]))
```
Comment (shmpanski) [inline scripts/evaluate_model.py]: Переменная где-то определена?

File: scripts/evaluate_model.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            context = batch["context"]
+            candidates = batch["candidates"]
+
+            inputs = tokenizer(
+                context,
+                return_tensors="pt",
+                max_length=128,
+                truncation=True,
+                padding=True,
+            ).to(device)
```
Comment (shmpanski) [inline scripts/evaluate_model.py]: Если предполагается использовать tokenizers только из huggingface, то имеет смысл занести токенизацию в `__getitem__` или сделать кастомную `collate_fn`

File: scripts/evaluate_model.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        history["bleu"].append(bleus[0])
+
+
+def main():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("-m", "--model")
+    parser.add_argument("-b", "--batch_size", type=int, default=64)
+
+    args = parser.parse_args()
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Let's change structure a little bit, so `main` method would be more reusable, e.g. import from another script or notebook.

I prefer to make
- `def configure_arg_parser() -> ArgumentParser` -- function to create argument parser w/ all arguments and their description
- `def main(...)` -- main method accept all necessary args, i.e. model name and batch size
After make entrypoint would be like that:
```python
if __name__ == "__main__":
    _args = configure_arg_parser().parse_args()
    main(**vars(_args))  # or main(_args.model, _args.batch_size) if there is no many arguments
```

File: scripts/evaluate_model.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        history["hit"].append(int(np.array(bleus).argmax() == 0))
+        history["bleu"].append(bleus[0])
+
+
+def main():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("-m", "--model")
+    parser.add_argument("-b", "--batch_size", type=int, default=64)
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Don't forget to add help, this will help you to write less documentation

File: scripts/evaluate_model.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    args = parser.parse_args()
+
+    convai2 = load_dataset("conv_ai_2")
+    dataset = ConvAI2CandidatesDataset(convai2)
+    dataloader = DataLoader(dataset, batch_size=args.batch_size)
+
+    tokenizer = AutoTokenizer.from_pretrained(args.model, truncation_side="left", padding_side="left")
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Truncation and padding sides are model specific, why are you hard-coded this values?

File: scripts/evaluate_model.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    device = "cuda" if torch.cuda_is_available() else "cpu"
+
+    history = {"bleu": [], "hit": []}
+
+    model = model.to(device)
+
+    model.eval()
+    with torch.no_grad():
+        for batch in (pbar := tqdm(dataloader)):
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: I think this line is not pythonic. They add this functionality in 3.10, but I it reduce simplicity of code 😅
I suggest to define progress bar in separate expression.

File: scripts/evaluate_model.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            pbar.set_description("Calculating metrics")
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: You can set description in progress bar initialisation.

File: scripts/evaluate_model.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            context = batch["context"]
+            candidates = batch["candidates"]
+
+            inputs = tokenizer(
+                context,
+                return_tensors="pt",
+                max_length=128,
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: This should be either argument to pass or extracted from the model. Something like minimum between model maximum len and what use pass.

Or if you don't want to pass len with arguments, you can define global constant, so we can easily change it in future.

File: scripts/evaluate_model.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                truncation=True,
+                padding=True,
+            ).to(device)
+            reply_ids = model.generate(**inputs, max_length=64)
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Generate has a lot of arguments to tune, for now you are using greedy decoding. But there is an option to use beam search or different sampling strategies. Let's define a global dict with parameters for this function. Most important: `max_length`, `do_sample`, `num_beams`, `top_k`, `top_p`, `repetition_penalty`

File: scripts/evaluate_model.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        Args:
+            dataset (DatasetDict): ConvAI2 dataset from huggingface
+        """
+
+        np.random.seed(42)
+        candidates_dataset = {"context": [], "candidates": []}
+        ids = list(range(len(dataset["train"])))
+
+        for i, dialog in (pbar := tqdm(enumerate(dataset["train"]))):
+            pbar.set_description("Creating dataset")
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Same comments about progress bar

File: scripts/evaluate_model.py (hunk 2 above)
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Let's make seed global constant at top of the script.
```python
SEED = 42
```

File: scripts/evaluate_model.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+                        length = len(other_dialog)
+                        text = ""
+                        sender = "Human"
+
+                        while sender == "Human":
+                            j = np.random.randint(length)
+                            text = other_dialog[j]["text"]
+                            sender = other_dialog[j]["sender"]
+                        candidates.append(text)
+
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Have some questions here:
- Negative candidates never will be from the same dialog. Is it okay?
- Negative candidates always bot messages. Is it also okay?

I also suggest you to simplify this logic: let's create an array with all possible negative candidates and the just sample from it.

File: scripts/evaluate_model.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+            context = ""
+
+            for utterance in dialog["dialog"]:
+                if utterance["sender_class"] == "Bot" and context:
+                    candidates_dataset["context"].append(context)
+                    candidates = [utterance["text"]]
+                    dialog_ids = [i]
+
+                    while i in dialog_ids:
+                        dialog_ids = np.random.choice(ids, size=19, replace=False)
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: This `19` should dataset parameter, don't hard-code values

File: scripts/evaluate_model.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        Args:
+            dataset (DatasetDict): ConvAI2 dataset from huggingface
+        """
+
+        np.random.seed(42)
+        candidates_dataset = {"context": [], "candidates": []}
+        ids = list(range(len(dataset["train"])))
+
+        for i, dialog in (pbar := tqdm(enumerate(dataset["train"]))):
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: This cycle is too complex, let's split it into a few functions. For example, extracting samples from one dialog may be moved into a separate function.

**Review by SpirinEgor — CHANGES_REQUESTED** [review]

File: scripts/evaluation_utils.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+import numpy as np
+
+import torch
+from torch import nn
+from torch.utils.data import Dataset
+
+from tqdm import tqdm
+
+
+def configure_arg_parser():
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: Move this function to `scripts/evaluate_model.py`
1. You will never reuse it, this argument parser is specific only for one script
2. This function also describes evaluating parameters, so it is better to keep it in the main script. This way anyone can quickly check what and why they should specify.

File: scripts/evaluation_utils.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    parser.add_argument("-cl", "--max_context", type=int, default=256, help="Max length of context fed into model")
+    parser.add_argument("-gl", "--max_gen", type=int, default=64, help="Max length of generated output")
+    parser.add_argument("-d", "--device", default="cpu", help="Device on which to evaluate the model")
+    parser.add_argument("-ds", "--do_sample", default=True, help="parameter for generation method")
+    parser.add_argument("-nb", "--num_beams", default=4, help="Parameter for generation method")
+
+    return parser
+
+
+class ConvAI2Dataset(Dataset):
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: This class you will use during model training. It is the main class for your data. So this should be in your sources. For example, in `src/data/dataset.py`

File: scripts/evaluation_utils.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+    """ConvAI2 Dataset"""
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: This docstring is kind of useless. Either add a dataset description or remove it.

File: scripts/evaluation_utils.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+    def __init__(self, path):
+        """
+        Args:
+            path (str): File path to ConvAI2 dataset
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: If you know that `path` has a `str` type then it is better to specify it in a pythonic way:
```python
def __init__(self, path: str):
```

Comment (SpirinEgor) [reply scripts/evaluation_utils.py]: All in all, it is better to write type hints whenever it is possible

File: scripts/evaluation_utils.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+        """
+
+        persona = "your persona"
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: Let's move to class constant, `_PERSONA_PREFIX`

File: scripts/evaluation_utils.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+        context = ""
+
+        candidates = []
+
+        dataset = {"context": [], "candidates": []}
+
+        with open(path, "r") as f:
+            for line in f.readlines():
+                if line[2 : 2 + len(persona)] == persona:
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: `if line[2:].startswith(self._PERSONA_PREFIX)`

File: scripts/evaluation_utils.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+                    utterance1, utterance2, _, candidates = line[2:].split("\t")
+                    candidates = candidates.split("|")[:-1]
+
+                    if prev_candidates:
+                        dataset["context"].append(context)
+                        dataset["candidates"].append([utterance1] + candidates)
+
+                    context += "<s>" + utterance1 + "</s>" + "<s>" + utterance2 + "</s>"
+
+        self.dataset = dataset
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: It's always good to log all information. So you are always aware of what happens in your code.
I suggest using [loguru](https://github.com/Delgan/loguru).

For example, in the dataset you can log some statistics at the end:
```python
from loguru import logger

...

logger.info(f"Init dataset w/ {len(dataset["context"]} dialog, # utterance {} ...")
```

File: scripts/evaluation_utils.py [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+    def __getitem__(self, idx):
+        sample = {
+            "context": self.dataset["context"][idx],
+            "candidates": self.dataset["candidates"][idx],
+        }
+
+        return sample
+
+
+def calculate_hits_ppl(model, tokenizer, dataset, max_len, device):
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: These functions also may be useful in model training, so let's move them into main sources. For example, `src/metrics.py`

File: scripts/evaluation_utils.py [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+    hits = []
+    ppl = []
+
+    loss_fn = nn.CrossEntropyLoss(reduction="none")
```
Comment (SpirinEgor) [inline scripts/evaluation_utils.py]: I think that there is no need for separate class, you can use [`torch.nn.functional.cross_entropy`](https://pytorch.org/docs/stable/generated/torch.nn.functional.cross_entropy.html#torch.nn.functional.cross_entropy)

**Review by SpirinEgor — CHANGES_REQUESTED** [review]

File: scripts/evaluate_model.py [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+
+from src.data.dataset import ConvAI2Dataset
+from src.metrics import calculate_f1
+from src.metrics import calculate_hits_ppl
+
+
+def configure_arg_parser():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("-m", "--diffusion", help="Seq2seq diffusion name from huggingface")
```
Comment (SpirinEgor) [inline scripts/evaluate_model.py]: Diffusion name?

File: src/metrics.py [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+from tqdm import tqdm
+
+
+def calculate_hits_ppl(model, tokenizer, dataset, max_len: int, device: str):
+    hits = []
+    ppl = []
+
+    model.eval()
+    with torch.no_grad():
```
Comment (SpirinEgor) [inline src/metrics.py]: Just add decorator `@torch.no_grad()` to this function.

File: src/metrics.py [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+    max_context_len,
+    max_gen_len,
+    do_sample,
+    num_beams,
+    device,
+):
+    f1 = []
+
+    model.eval()
+    with torch.no_grad():
```
Comment (SpirinEgor) [inline src/metrics.py]: And you can use a decorator here too. This will avoid unnecessary nesting.

File: src/metrics.py [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+                    * shift_attention_mask_batch
+                ).sum(1)
+                / shift_attention_mask_batch.sum(1)
+            ).view(-1, num_candidates)
+
+            hits.append(perplexity_batch.argmin(dim=1).eq(0))
+            ppls.append(perplexity_batch[:, 0])
+
+    hit = torch.concat(hits).float().mean().item()
+    ppl = torch.concat(ppls).float().median().item()
```
Comment (SpirinEgor) [inline src/metrics.py]: Why median for the perplexity?

**Review by SpirinEgor — COMMENTED** [review]
🔥
